### PR TITLE
0.7compat

### PR DESCRIPTION
**Pull request by gdkrmr**
I gave it a shot but cannot find out why I get the following error:

https://travis-ci.org/gdkrmr/JLD2.jl/jobs/412170446#L501

```
ERROR: LoadError: LoadError: UndefVarError: arr_undef not defined
```
This gets a warning in 0.7:
```
┌ Warning: Deprecated syntax `implicit assignment to global variable `arr_undef``.
│ Use `global arr_undef` instead.
└ @ none:0
```

I removed all the deprecation warning I could locate. I am not sure about about the right usage of `unsafe_wrap` instead of `Vector{UInt8}`, you may want to check those.

I also don't know how to deal with this one:

https://travis-ci.org/gdkrmr/JLD2.jl/jobs/412170446#L495
```
┌ Warning: Deprecated syntax `"begin" inside indexing expression` at /home/travis/build/gdkrmr/JLD2.jl/src/data.jl:900.
└ @ ~/build/gdkrmr/JLD2.jl/src/data.jl:900
```

### CODE REVIEW

File: src/data.jl [hunk 1]
```diff
@@ -761,7 +761,7 @@ function h5convert!(out::Pointers, fls::FixedLengthString, f::JLDFile, x, ::JLDW
     (unsafe_copyto!(convert(Ptr{UInt8}, out), pointer(x), fls.length); nothing)
 end
 h5convert!(out::Pointers, ::Type{Vlen{String}}, f::JLDFile, x, wsession::JLDWriteSession) =
-    store_vlen!(out, UInt8, f, Vector{UInt8}(x), wsession)
+    store_vlen!(out, UInt8, f, unsafe_wrap(Array{UInt8}, pointer(x), ncodeunits(x)), wsession)
```
Comment (JeffBezanson) [inline src/data.jl]: This is incorrect, as it can drop references to `x`. If `x` is a string, we want
```
unsafe_wrap(Vector{UInt8}, x)
```

**Comment by JeffBezanson** [comment]
Also needs:
```
--- a/src/data.jl
+++ b/src/data.jl
@@ -894,7 +894,7 @@ function jlconvert(::ReadRepresentation{T,DataTypeODR()}, f:
     if hasparams
         paramrefs = jlconvert(ReadRepresentation{RelOffset,Vlen{RelOffset}}(), 
                               ptr+odr_sizeof(Vlen{UInt8}), NULL_REFERENCE)
-        params = Any[begin
+        params = Any[let
             # If the reference is to a committed datatype, read the datatype
             nulldt = CommittedDatatype(UNDEFINED_ADDRESS, 0)
             cdt = get(f.datatype_locations, ref, nulldt)
```

**Comment by JeffBezanson** [comment]
Also
```
--- a/test/rw.jl
+++ b/test/rw.jl
@@ -517,7 +517,7 @@ for ioty in [JLD2.MmapIO, IOStream], compress in [false, tru
     @check fidr typevar_lb_ub
 
     # Special cases for reading undefs
-    arr_undef = read(fidr, "arr_undef")
+    global arr_undef = read(fidr, "arr_undef")
     if !isa(arr_undef, Array{Any, 1}) || length(arr_undef) != 1 || isassigned(a
         error("For arr_undef, read value does not agree with written value")
     end
```

Hopefully that will fix the remaining issues.

**Comment by gdkrmr** [comment]
so instead of `[begin ... end for i in I]` you can always do `[let ... end for i in I]`? When I was searching the documentation I was not able to find this.

**Pull request merged**